sections/section_util.go

```diff
@@ -0,0 +1,382 @@
+}
+
+type USPUTCoreSegment struct {
+	Version                             byte
+	SharingNotice                       byte
+	SaleOptOutNotice                    byte
+	TargetedAdvertisingOptOutNotice     byte
+	SensitiveDataProcessingOptOutNotice byte
+	SaleOptOut                          byte
+	TargetedAdvertisingOptOut           byte
+	SensitiveDataProcessing             []byte
+	KnownChildSensitiveDataConsents     byte
+	MspaCoveredTransaction              byte
+	MspaOptOutOptionMode                byte
+	MspaServiceProviderMode             byte
+}
+
+type USPNATCoreSegment struct {
+	Version                             byte
+	SharingNotice                       byte
+	SaleOptOutNotice                    byte
+	SharingOptOutNotice                 byte
+	TargetedAdvertisingOptOutNotice     byte
+	SensitiveDataProcessingOptOutNotice byte
+	SensitiveDataLimitUseNotice         byte
+	SaleOptOut                          byte
+	SharingOptOut                       byte
+	TargetedAdvertisingOptOut           byte
+	SensitiveDataProcessing             []byte
+	KnownChildSensitiveDataConsents     []byte
+	PersonalDataConsents                byte
+	MspaCoveredTransaction              byte
+	MspaOptOutOptionMode                byte
+	MspaServiceProviderMode             byte
+}
+
+func errorHelper(name string, err error) error {
+	return fmt.Errorf("unable to set field %s due to parse error: %s", name, err.Error())
+}
+
+func NewCommonUSCoreSegment(sensitiveDataFields int, knownChildDataFields int, bs *util.BitStream) (CommonUSCoreSegment, error) {
+	var commonUSCore CommonUSCoreSegment
+	var err error
+
+	commonUSCore.Version, err = bs.ReadByte6()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.Version", err)
+	}
+
+	commonUSCore.SharingNotice, err = bs.ReadByte2()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.SharingNotice", err)
+	}
+
+	commonUSCore.SaleOptOutNotice, err = bs.ReadByte2()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.SaleOptOutNotice", err)
+	}
+
+	commonUSCore.TargetedAdvertisingOptOutNotice, err = bs.ReadByte2()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.TargetedAdvertisingOptOutNotice", err)
+	}
+
+	commonUSCore.SaleOptOut, err = bs.ReadByte2()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.SaleOptOut", err)
+	}
+
+	commonUSCore.TargetedAdvertisingOptOut, err = bs.ReadByte2()
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.TargetedAdvertisingOptOut", err)
+	}
+
+	commonUSCore.SensitiveDataProcessing, err = bs.ReadTwoBitField(sensitiveDataFields)
+	if err != nil {
+		return commonUSCore, errorHelper("CoreSegment.SensitiveDataProcessing", err)
+	}
+
+	if knownChildDataFields > 0 {
+		commonUSCore.KnownChildSensitiveDataConsentsArr, err = bs.ReadTwoBitField(knownChildDataFields)
+		if err != nil {
+			return commonUSCore, errorHelper("CoreSegment.KnownChildSensitiveDataConsentsArr", err)
+		}
+	} else {
```

Comment:
Isn't this `else` the same as `commonUSCore.KnownChildSensitiveDataConsentsArr, err = bs.ReadTwoBitField(1)`. It might be clearer to use 1 rather than zero so the reader knows there is one field/result in there.

Reply:
I tried to simplify this by just making it an array. There are sections that use it as an array and some that use it as an int. I thought if it's kept an array, the caller can use the single entry as the int value. Let me know if this is a better approach or if I should keep the values separated.

Reply:
Gotcha! I think in that case, it seems more straight forward to have it just be an array. If the caller needs to grab the int value, it can just be accessed in the array. Otherwise, this keeps this block more clear I think.